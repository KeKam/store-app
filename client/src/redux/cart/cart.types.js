const CartActionTypes = {
  TOGGLE_CART: 'TOGGLE_CART',
  ADD_TO_CART: 'ADD_TO_CART',
  REMOVE_FROM_CART: 'REMOVE_FROM_CART',
  REMOVE_ITEM: 'REMOVE_ITEM',
  CLEAR_CART: 'CLEAR_CART',
  SET_CART_FROM_FIREBASE: 'SET_CART_FROM_FIREBASE'
};

export default CartActionTypes;
