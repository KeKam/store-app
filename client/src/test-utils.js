import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

import rootReducer from './redux/root-reducer';

// const render = (
//   component,
//   {
//     initialState,
//     store = createStore(rootReducer, initialState),
//     ...renderOptions
//   } = {}
// ) => {
//   return {
//     ...rtlRender(<Provider store={store}>{component}</Provider>),
//     store,
//   };
//   // const Wrapper = ({ children }) => {
//   //   return <Provider store={store}>{children}</Provider>;
//   // };
//   // return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// };

const render = (component, { initialState } = {}) => {
  const actions = [];
  const observerMiddleware = () => (next) => (action) => {
    actions.push(action);
    return next(action);
  };
  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(observerMiddleware)
  );
  const utils = {
    dispatch(action) {
      return store.dispatch(action);
    },
    getDispatchedActions() {
      return actions;
    },
    getState() {
      return store.getState();
    },
  };
  return {
    ...rtlRender(<Provider store={store}>{component}</Provider>),
    ...utils,
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
