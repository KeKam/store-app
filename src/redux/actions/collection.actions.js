import { CollectionActionTypes } from '../types/collection.types';
import {
  firestore,
  convertCollectionSnapshotToMap
} from '../../firebase/firebase';

export const startFetchCollection = () => ({
  type: CollectionActionTypes.START_FETCH_COLLECTION
});

export const fetchCollectionSuccess = collectionMap => ({
  type: CollectionActionTypes.FETCH_COLLECTION_SUCCESS,
  payload: collectionMap
});

export const fetchCollectionFailure = error => ({
  type: CollectionActionTypes.FETCH_COLLECTION_FAILURE,
  payload: error
});

export const startAsyncFetchCollection = () => {
  return async dispatch => {
    const collectionRef = firestore.collection('collection');
    dispatch(startFetchCollection());

    try {
      const response = await collectionRef.get();
      const collectionMap = convertCollectionSnapshotToMap(response);
      dispatch(fetchCollectionSuccess(collectionMap));
    } catch (error) {
      dispatch(fetchCollectionFailure(error));
    }
  };
};
