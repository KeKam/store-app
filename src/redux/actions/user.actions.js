export const SET_USER = user => {
  return {
    type: 'SET_USER',
    payload: user
  };
};
