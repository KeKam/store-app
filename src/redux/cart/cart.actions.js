import CartActionTypes from './cart.types';

export const toggleCart = () => ({
  type: CartActionTypes.TOGGLE_CART
});

export const addToCart = item => ({
  type: CartActionTypes.ADD_TO_CART,
  payload: item
});

export const removeFromCart = item => ({
  type: CartActionTypes.REMOVE_FROM_CART,
  payload: item
});

export const removeItem = item => ({
  type: CartActionTypes.REMOVE_ITEM,
  payload: item
});

export const clearCart = () => ({
  type: CartActionTypes.CLEAR_CART
});

export const setCartFromFirebase = cartItems => ({
  type: CartActionTypes.SET_CART_FROM_FIREBASE,
  payload: cartItems
});
