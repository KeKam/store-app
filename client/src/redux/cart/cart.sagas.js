import { takeLatest, put, call, all, select } from 'redux-saga/effects';

import UserActionTypes from '../user/user.types';
import CartActionTypes from './cart.types';
import { getCurrentUserCart } from '../../firebase/firebase';
import { selectCurrentUser } from '../user/user.selectors';
import { selectCartItems, selectCartHidden } from './cart.selectors';
import { clearCart, setCartFromFirebase, toggleCart } from './cart.actions';

export function* clearCartOnSignOut() {
  const hidden = yield select(selectCartHidden);
  yield put(clearCart());

  if (!hidden) {
    yield put(toggleCart());
  }
}

export function* updateCartInFirebase() {
  const currentUser = yield select(selectCurrentUser);

  if (currentUser) {
    try {
      const cartRef = yield getCurrentUserCart(currentUser.id);
      const cartItems = yield select(selectCartItems);
      yield cartRef.update({ cartItems });
    } catch (error) {
      console.log(error);
    }
  }
}

export function* checkCartFromFirebase({ payload: user }) {
  const cartRef = yield getCurrentUserCart(user.id);
  const cartSnapshot = yield cartRef.get();
  yield put(setCartFromFirebase(cartSnapshot.data().cartItems));
}

export function* onSignInSuccess() {
  yield takeLatest(UserActionTypes.SIGN_IN_SUCCESS, checkCartFromFirebase);
}

export function* onSignOutSuccess() {
  yield takeLatest(UserActionTypes.SIGN_OUT_SUCCESS, clearCartOnSignOut);
}

export function* onCartChange() {
  yield takeLatest(
    [
      CartActionTypes.ADD_TO_CART,
      CartActionTypes.REMOVE_FROM_CART,
      CartActionTypes.CLEAR_CART,
    ],
    updateCartInFirebase
  );
}

export function* cartSagas() {
  yield all([
    call(onSignInSuccess),
    call(onSignOutSuccess),
    call(onCartChange),
  ]);
}
