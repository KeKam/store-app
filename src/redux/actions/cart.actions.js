import { CartActionTypes } from '../types/cart.types';

export const toggleCart = () => {
  return {
    type: CartActionTypes.TOGGLE_CART
  };
};

export const addToCart = item => {
  return {
    type: CartActionTypes.ADD_TO_CART,
    payload: item
  };
};

export const removeFromCart = item => {
  return {
    type: CartActionTypes.REMOVE_FROM_CART,
    payload: item
  };
};

export const removeItem = item => {
  return {
    type: CartActionTypes.REMOVE_ITEM,
    payload: item
  };
};
