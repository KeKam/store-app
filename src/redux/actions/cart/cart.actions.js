import { CartActionTypes } from '../../types/cart.types';

export const toggleCart = () => {
  return {
    type: CartActionTypes.TOGGLE_CART
  };
};
