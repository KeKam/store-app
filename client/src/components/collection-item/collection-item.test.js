import React from 'react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import { render, screen } from '../../test-utils';
import { addToCart } from '../../redux/cart/cart.actions';
import CollectionItem from '../collection-item/collection-item';

const history = createMemoryHistory();

describe('CollectionItem component', () => {
  let mockItem;
  let mockState;

  beforeEach(() => {
    mockItem = {
      name: 'Winter falls',
      imageUrl: 'www.testImage.com',
      price: 100,
    };

    mockState = {
      currentUser: {
        email: 'testUser@gmail.com',
        password: 'testUser123',
      },
    };
  });

  it('should render CollectionItem component', () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Router history={history}>
        <CollectionItem item={mockItem} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByTestId('image')).toBeInTheDocument();
    expect(screen.getByTestId('title')).toHaveTextContent('Winter fall');
    expect(screen.getByTestId('price')).toHaveTextContent('100 â‚¬');
    expect(screen.getByTestId('link')).toHaveTextContent('Sign in required');
    expect(screen.queryByTestId('button')).not.toBeInTheDocument();
  });

  it('should trigger history push on click', () => {
    render(
      <Router history={history}>
        <CollectionItem item={mockItem} />
      </Router>
    );

    history.push = jest.fn();

    userEvent.click(screen.getByText('Sign in required'));

    expect(history.push).toHaveBeenCalledTimes(1);
    expect(history.push).toHaveBeenCalledWith('/login');
  });

  it('should show add to cart button if there is a user', () => {
    render(
      <Router history={history}>
        <CollectionItem item={mockItem} />
      </Router>,
      { initialState: { user: mockState } }
    );

    expect(screen.getByTestId('button')).toHaveTextContent('ADD TO CART');
    expect(screen.queryByTestId('link')).not.toBeInTheDocument();
  });

  it('should dispatch addToCart action on click', () => {
    const { getDispatchedActions } = render(
      <Router history={history}>
        <CollectionItem item={mockItem} />
      </Router>,
      { initialState: { user: mockState } }
    );

    userEvent.click(screen.getByTestId('button'));

    expect(getDispatchedActions()).toContainEqual(addToCart(mockItem));
  });
});
