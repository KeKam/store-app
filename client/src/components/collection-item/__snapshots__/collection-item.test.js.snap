// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionItem component should render CollectionItem component 1`] = `
<DocumentFragment>
  <div
    class="sc-bwzfXH srITS"
  >
    <div
      class="sc-htpNat dTlvyq"
      data-testid="image"
    />
    <div
      class="sc-bxivhb oDKzY"
    >
      <span
        class="sc-EHOje ckFQmK"
        data-testid="title"
      >
        Winter falls
      </span>
      <span
        class="sc-bZQynM bjXVPL"
        data-testid="price"
      >
        100 â‚¬
      </span>
    </div>
    <a
      class="sc-gzVnrw jnQMmI"
      data-testid="link"
      href="/login"
    >
      <h2
        class="sc-htoDjs bRjjzo"
      >
        Sign in required
      </h2>
    </a>
  </div>
</DocumentFragment>
`;
